python/sparkdl/graph/spark_utils.py

```diff
@@ -0,0 +1,108 @@
+#
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+
+import tensorframes as tfs
+
+import sparkdl.graph.utils as tfx
+from sparkdl.utils import jvmapi as JVMAPI
+
+logger = logging.getLogger('sparkdl')
+
+def makeGraphUDF(graph, udf_name, fetches, feeds_to_fields_map=None, blocked=False, register=True):
+    """
+    Create a Spark SQL UserDefinedFunction from a given TensorFlow Graph
+
+    The following example creates a UDF that takes the input
+    from a DataFrame column named 'image_col' and produce some random prediction.
+
+    .. code-block:: python
+
+        from sparkdl.graph.spark_utils import makeUDF
+        
+        with IsolatedSession() as issn:
+            x = tf.placeholder(tf.double, shape=[], name="input_x")
+            z = tf.add(x, 3, name='z')
+            makeGraphUDF(issn.graph, "my_tensorflow_udf", [z])
+
+    Then this function can be used in a SQL query.
+
+    .. code-block:: python
+
+        df = spark.createDataFrame([Row(xCol=float(x)) for x in range(100)])
+        df.createOrReplaceTempView("my_float_table")
+        spark.sql("select my_tensorflow_udf(xCol) as zCol from my_float_table").show()            
+
+    :param graph: :py:class:`tf.Graph`, a TensorFlow Graph
+    :param udf_name: str, name of the SQL UDF
+    :param fetches: list, output tensors of the graph
+    :param feeds_to_fields_map: a dict of str -> str,
+                                The key is the name of a placeholder in the current
+                                TensorFlow graph of computation.
+                                The value is the name of a column in the dataframe.
+                                For now, only the top-level fields in a dataframe are supported.
+
+                                .. note:: For any placeholder that is
+                                          not specified in the feed dictionary,
+                                          the name of the input column is assumed to be
+                                          the same as that of the placeholder.
+
+    :param blocked: bool, if set to True, the TensorFrames will execute the function
+                    over blocks/batches of rows. This should provide better performance.
+                    Otherwise, the function is applied to individual rows
```

Comment:
Ah yeah, for the Keras udf registration functinon we do need to talk about `SpImage`. It might fail if the images have different widths and heights, but it could succeed if they all had the same widths and heights right?